File: jebat-sdk-ts/src/websocket.ts
/**
 * WebSocket client for JEBAT real-time streaming.
 * Provides auto-reconnect, message queuing, and event callbacks.
 */

import WebSocket from 'ws';
import type { ConnectionState, StreamChunkType } from './types';

/**
 * Stream chunk from WebSocket.
 */
export interface StreamChunk {
  type: StreamChunkType;
  content: string;
  data?: unknown;

  isFinal(): boolean;
}

const StreamChunkImpl = class implements StreamChunk {
  constructor(
    public readonly type: StreamChunkType,
    public readonly content: string,
    public readonly data?: unknown
  ) {}

  isFinal(): boolean {
    return this.type === 'complete' || this.type === 'error';
  }
};

/**
 * WebSocket message structure.
 */
interface WSMessage {
  type: string;
  content?: string;
  data?: unknown;
  [key: string]: unknown;
}

/**
 * Event callback types.
 */
type ConnectCallback = () => void | Promise<void>;
type DisconnectCallback = () => void | Promise<void>;
type MessageCallback = (data: WSMessage) => void | Promise<void>;
type ErrorCallback = (error: Error) => void | Promise<void>;

/**
 * JEBAT WebSocket client for real-time streaming.
 *
 * @example
 * ```typescript
 * const ws = new JebatWebSocketClient('ws://localhost:8000/ws/chat', accessToken);
 * await ws.connect();
 *
 * ws.on('message', (data) => console.log('Received:', data));
 *
 * for await (const chunk of ws.stream({ message: 'Hello!', mode: 'deep' })) {
 *   process.stdout.write(chunk.content);
 * }
 * ```
 */
export class JebatWebSocketClient {
  private url: string;
  private token: string;
  private ws: WebSocket | null = null;
  private state: ConnectionState = 'disconnected';
  private autoReconnect: boolean;
  private maxReconnectAttempts: number;
  private reconnectDelay: number;
  private reconnectAttempts: number = 0;
  private messageQueue: WSMessage[] = [];
  private resolveQueue: Array<(value: WSMessage) => void> = [];
  private receiveTask: Promise<void> | null = null;

  // Event callbacks
  private callbacks: {
    connect: ConnectCallback[];
    disconnect: DisconnectCallback[];
    message: MessageCallback[];
    error: ErrorCallback[];
  } = {
    connect: [],
    disconnect: [],
    message: [],
    error: [],
  };

  constructor(
    url: string,
    token: string,
    options: {
      autoReconnect?: boolean;
      maxReconnectAttempts?: number;
      reconnectDelay?: number;
      pingInterval?: number;
    } = {}
  ) {
    this.url = url;
    this.token = token;
    this.autoReconnect = options.autoReconnect ?? true;
    this.maxReconnectAttempts = options.maxReconnectAttempts ?? 5;
    this.reconnectDelay = options.reconnectDelay ?? 1000;
  }

  get connectionState(): ConnectionState {
    return this.state;
  }

  get isConnected(): boolean {
    return this.state === 'connected' && this.ws?.readyState === WebSocket.OPEN;
  }

  /**
   * Register event callback.
   */
  on(event: 'connect', callback: ConnectCallback): () => void;
  on(event: 'disconnect', callback: DisconnectCallback): () => void;
  on(event: 'message', callback: MessageCallback): () => void;
  on(event: 'error', callback: ErrorCallback): () => void;
  on(event: string, callback: Function): () => void {
    const callbacks = this.callbacks[event as keyof typeof this.callbacks];
    if (callbacks) {
      callbacks.push(callback as any);
      return () => {
        const idx = callbacks.indexOf(callback as any);
        if (idx > -1) callbacks.splice(idx, 1);
      };
    }
    throw new Error(`Unknown event: ${event}`);
  }

  /**
   * Connect to WebSocket server.
   */
  async connect(): Promise<void> {
    if (this.isConnected) return;

    this.setState('connecting');

    return new Promise((resolve, reject) => {
      const headers = { Authorization: `Bearer ${this.token}` };
      this.ws = new WebSocket(this.url, { headers });

      this.ws.onopen = () => {
        this.setState('connected');
        this.reconnectAttempts = 0;
        this.startReceiveLoop();
        this.trigger('connect');
        resolve();
      };

      this.ws.onerror = (error) => {
        if (this.state === 'connecting') {
          reject(new Error(`Connection failed: ${error}`));
        }
        this.trigger('error', new Error(`WebSocket error: ${error}`));
      };

      this.ws.onclose = async (event) => {
        this.setState('disconnected');
        this.trigger('disconnect');

        if (this.autoReconnect && this.state !== 'closed') {
          await this.handleReconnect();
        }
      };
    });
  }

  /**
   * Disconnect from WebSocket server.
   */
  async disconnect(): Promise<void> {
    this.autoReconnect = false;
    this.setState('closed');

    if (this.receiveTask) {
      // The receive loop will exit on close
    }

    if (this.ws) {
      this.ws.close(1000, 'Client disconnect');
      this.ws = null;
    }
  }

  /**
   * Send a message.
   */
  send(data: WSMessage): void {
    if (!this.isConnected) {
      throw new Error('Not connected');
    }
    this.ws!.send(JSON.stringify(data));
  }

  /**
   * Send a chat message.
   */
  sendChat(message: string, mode: string = 'deliberate'): void {
    this.send({
      type: 'chat',
      data: { message, mode, stream: true },
    });
  }

  /**
   * Stream chat response.
   */
  async *stream(chatRequest: { message: string; mode: string }): AsyncGenerator<StreamChunk, void, unknown> {
    if (!this.isConnected) {
      await this.connect();
    }

    // Send chat request
    this.sendChat(chatRequest.message, chatRequest.mode);

    // Yield chunks from queue
    while (true) {
      try {
        // Wait for next message with timeout
        const data = await Promise.race([
          this.waitForMessage(),
          new Promise<never>((_, reject) =>
            setTimeout(() => reject(new Error('Stream timeout')), 30000)
          ),
        ]);

        const chunk = new StreamChunkImpl(
          data.type as StreamChunkType || 'content',
          data.content || '',
          data.data
        );

        yield chunk;

        if (chunk.isFinal()) {
          break;
        }
      } catch (error) {
        yield new StreamChunkImpl('error', error instanceof Error ? error.message : 'Stream error');
        break;
      }
    }
  }

  /**
   * Wait for next message from queue.
   */
  private waitForMessage(): Promise<WSMessage> {
    if (this.messageQueue.length > 0) {
      return Promise.resolve(this.messageQueue.shift()!);
    }

    return new Promise((resolve) => {
      this.resolveQueue.push(resolve);
    });
  }

  /**
   * Start background receive loop.
   */
  private startReceiveLoop(): void {
    if (!this.ws) return;

    this.receiveTask = (async () => {
      try {
        for await (const message of this.iterateMessages()) {
          try {
            const data = JSON.parse(message.toString()) as WSMessage;
            this.messageQueue.push(data);

            // Resolve waiting promises
            if (this.resolveQueue.length > 0) {
              const resolve = this.resolveQueue.shift()!;
              resolve(data);
            }

            this.trigger('message', data);
          } catch (error) {
            console.warn('Failed to parse WebSocket message:', error);
          }
        }
      } catch (error) {
        if (this.state !== 'closed') {
          this.trigger('error', error instanceof Error ? error : new Error('Receive loop error'));
        }
      }
    })();
  }

  /**
   * Async iterator for WebSocket messages.
   */
  private async *iterateMessages(): AsyncGenerator<Buffer, void, unknown> {
    if (!this.ws) return;

    while (this.ws.readyState === WebSocket.OPEN) {
      yield await new Promise<Buffer>((resolve, reject) => {
        this.ws!.once('message', (data: Buffer) => resolve(data));
        this.ws!.once('error', reject);
        this.ws!.once('close', () => reject(new Error('Connection closed')));
      });
    }
  }

  /**
   * Handle automatic reconnection.
   */
  private async handleReconnect(): Promise<void> {
    if (this.state === 'closed') return;

    this.setState('reconnecting');

    while (this.reconnectAttempts < this.maxReconnectAttempts) {
      this.reconnectAttempts++;
      const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1);

      console.log(`[JEBAT SDK] Reconnecting in ${delay}ms (attempt ${this.reconnectAttempts})`);
      await new Promise(r => setTimeout(r, delay));

      try {
        await this.connect();
        return;
      } catch (error) {
        console.warn(`[JEBAT SDK] Reconnect failed:`, error);
      }
    }

    console.error('[JEBAT SDK] Max reconnect attempts reached');
    this.setState('disconnected');
    this.trigger('error', new Error('Max reconnect attempts reached'));
  }

  /**
   * Update connection state and trigger callbacks.
   */
  private setState(state: ConnectionState): void {
    this.state = state;
  }

  /**
   * Trigger event callbacks.
   */
  private async trigger(event: 'connect' | 'disconnect' | 'message' | 'error', data?: any): Promise<void> {
    const callbacks = this.callbacks[event];
    for (const callback of callbacks) {
      try {
        await callback(data);
      } catch (error) {
        console.error(`[JEBAT SDK] Callback error (${event}):`, error);
      }
    }
  }
}

File: jebat_sdk/src/jebat_sdk/websocket.py
"""
WebSocket client for JEBAT SDK streaming.

Provides WebSocket support for real-time chat streaming.
"""

import asyncio
import json
from typing import AsyncGenerator, Dict, Any, Optional, Callable, Awaitable
from dataclasses import dataclass, field
from enum import Enum
import logging

import websockets
from websockets.client import WebSocketClientProtocol

logger = logging.getLogger(__name__)


class ConnectionState(Enum):
    """WebSocket connection state."""
    DISCONNECTED = "disconnected"
    CONNECTING = "connecting"
    CONNECTED = "connected"
    RECONNECTING = "reconnecting"
    CLOSED = "closed"


@dataclass
class StreamChunk:
    """Stream chunk from WebSocket."""
    type: str  # "content", "thinking", "complete", "error"
    content: str = ""
    data: Optional[Dict[str, Any]] = None

    def is_final(self) -> bool:
        return self.type in ("complete", "error")


class JebatWebSocketClient:
    """
    WebSocket client for JEBAT streaming API.

    Usage:
        async with JebatWebSocketClient("ws://localhost:8000/ws/chat", token) as ws:
            async for chunk in ws.stream(chat_request):
                print(chunk.content, end="", flush=True)
    """

    def __init__(
        self,
        url: str,
        token: str,
        auto_reconnect: bool = True,
        max_reconnect_attempts: int = 5,
        reconnect_delay: float = 1.0,
        ping_interval: float = 20.0,
        ping_timeout: float = 10.0,
    ):
        """
        Initialize WebSocket client.

        Args:
            url: WebSocket URL (e.g., ws://localhost:8000/ws/chat)
            token: JWT access token for authentication
            auto_reconnect: Whether to auto-reconnect on disconnect
            max_reconnect_attempts: Maximum reconnection attempts
            reconnect_delay: Initial delay between reconnects (seconds)
            ping_interval: Ping interval in seconds
            ping_timeout: Ping timeout in seconds
        """
        self.url = url
        self.token = token
        self.auto_reconnect = auto_reconnect
        self.max_reconnect_attempts = max_reconnect_attempts
        self.reconnect_delay = reconnect_delay
        self.ping_interval = ping_interval
        self.ping_timeout = ping_timeout

        self._ws: Optional[WebSocketClientProtocol] = None
        self._state = ConnectionState.DISCONNECTED
        self._reconnect_attempts = 0
        self._receive_task: Optional[asyncio.Task] = None
        self._message_queue: asyncio.Queue = asyncio.Queue()
        self._callbacks: Dict[str, list] = {
            "connect": [],
            "disconnect": [],
            "message": [],
            "error": [],
        }

    @property
    def state(self) -> ConnectionState:
        return self._state

    @property
    def is_connected(self) -> bool:
        return self._state == ConnectionState.CONNECTED

    def on(self, event: str, callback: Callable):
        """Register event callback."""
        if event in self._callbacks:
            self._callbacks[event].append(callback)

    async def connect(self) -> None:
        """Establish WebSocket connection."""
        if self._state == ConnectionState.CONNECTED:
            return

        self._state = ConnectionState.CONNECTING
        logger.info(f"Connecting to {self.url}")

        try:
            headers = {"Authorization": f"Bearer {self.token}"}
            self._ws = await websockets.connect(
                self.url,
                extra_headers=headers,
                ping_interval=self.ping_interval,
                ping_timeout=self.ping_timeout,
            )

            self._state = ConnectionState.CONNECTED
            self._reconnect_attempts = 0
            logger.info("WebSocket connected")

            # Start receive loop
            self._receive_task = asyncio.create_task(self._receive_loop())

            # Trigger connect callbacks
            for cb in self._callbacks["connect"]:
                await self._safe_call(cb)

        except Exception as e:
            self._state = ConnectionState.DISCONNECTED
            logger.error(f"Connection failed: {e}")
            await self._trigger_error(e)
            raise

    async def disconnect(self) -> None:
        """Disconnect WebSocket."""
        self._state = ConnectionState.CLOSED
        self.auto_reconnect = False

        if self._receive_task:
            self._receive_task.cancel()
            try:
                await self._receive_task
            except asyncio.CancelledError:
                pass

        if self._ws:
            await self._ws.close()
            self._ws = None

        logger.info("WebSocket disconnected")
        await self._trigger_disconnect()

    async def _receive_loop(self) -> None:
        """Background task to receive messages."""
        try:
            async for message in self._ws:
                try:
                    data = json.loads(message)
                    await self._message_queue.put(data)
                    await self._trigger_message(data)
                except json.JSONDecodeError:
                    logger.warning(f"Invalid JSON: {message}")
                except Exception as e:
                    logger.error(f"Message handling error: {e}")

        except websockets.exceptions.ConnectionClosed:
            logger.info("WebSocket connection closed")
        except asyncio.CancelledError:
            pass
        except Exception as e:
            logger.error(f"Receive loop error: {e}")

        finally:
            if self.auto_reconnect and self._state != ConnectionState.CLOSED:
                await self._handle_reconnect()

    async def _handle_reconnect(self) -> None:
        """Handle automatic reconnection."""
        if self._state == ConnectionState.CLOSED:
            return

        self._state = ConnectionState.RECONNECTING

        while self._reconnect_attempts < 5:
            self._reconnect_attempts += 1
            delay = self.reconnect_delay * (2 ** (self._reconnect_attempts - 1))
            logger.info(f"Reconnecting in {delay}s (attempt {self._reconnect_attempts})")
            await asyncio.sleep(delay)

            try:
                await self.connect()
                return
            except Exception as e:
                logger.warning(f"Reconnect failed: {e}")

        logger.error("Max reconnect attempts reached")
        self._state = ConnectionState.DISCONNECTED
        await self._trigger_error(Exception("Max reconnect attempts reached"))

    async def send(self, data: Dict[str, Any]) -> None:
        """Send JSON message."""
        if not self._ws or not self.is_connected:
            raise ConnectionError("Not connected")
        await self._ws.send(json.dumps(data))

    async def stream(
        self,
        chat_request: Dict[str, Any],
    ) -> AsyncGenerator[StreamChunk, None]:
        """
        Stream chat response via WebSocket.

        Args:
            chat_request: Chat request dict with message, mode, etc.

        Yields:
            StreamChunk objects
        """
        if not self.is_connected:
            raise ConnectionError("Not connected")

        # Send chat request
        await self.send({"type": "chat", "data": chat_request})

        # Yield chunks from queue
        while True:
            try:
                # Wait for next message with timeout
                data = await asyncio.wait_for(
                    self._message_queue.get(),
                    timeout=30.0,
                )

                chunk = StreamChunk(
                    type=data.get("type", "content"),
                    content=data.get("content", ""),
                    data=data.get("data"),
                )
                yield chunk

                if chunk.is_final():
                    break

            except asyncio.TimeoutError:
                logger.warning("Stream timeout")
                yield StreamChunk(type="error", content="Stream timeout")
                break
            except Exception as e:
                logger.error(f"Stream error: {e}")
                yield StreamChunk(type="error", content=str(e))
                break

    async def send_chat(self, message: str, mode: str = "deliberate") -> None:
        """Send a chat message."""
        await self.send({
            "type": "chat",
            "data": {"message": message, "mode": mode, "stream": True}
        })

    # Event callbacks
    async def _safe_call(self, callback: Callable) -> None:
        """Safely call a callback."""
        try:
            result = callback()
            if asyncio.iscoroutine(result):
                await result
        except Exception as e:
            logger.error(f"Callback error: {e}")

    async def _trigger_message(self, data: Dict[str, Any]) -> None:
        for cb in self._callbacks["message"]:
            await self._safe_call(lambda: cb(data))

    async def _trigger_connect(self) -> None:
        for cb in self._callbacks["connect"]:
            await self._safe_call(cb)

    async def _trigger_disconnect(self) -> None:
        for cb in self._callbacks["disconnect"]:
            await self._safe_call(cb)

    async def _trigger_error(self, error: Exception) -> None:
        for cb in self._callbacks["error"]:
            await self._safe_call(lambda: cb(error))

    async def __aenter__(self) -> "JebatWebSocketClient":
        await self.connect()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        await self.disconnect()


class JebatWebSocketClientSync:
    """
    Synchronous wrapper for WebSocket client.

    Use this for synchronous code that needs streaming.
    """

    def __init__(self, *args, **kwargs):
        self._async_client = JebatWebSocketClient(*args, **kwargs)
        self._loop = None

    def _get_loop(self):
        try:
            return asyncio.get_running_loop()
        except RuntimeError:
            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)
            return loop

    def connect(self) -> None:
        loop = self._get_loop()
        loop.run_until_complete(self._async_client.connect())

    def disconnect(self) -> None:
        loop = self._get_loop()
        loop.run_until_complete(self._async_client.disconnect())

    def send_chat(self, message: str, mode: str = "deliberate") -> None:
        loop = self._get_loop()
        loop.run_until_complete(self._async_client.send_chat(message, mode))

    def stream_chat(
        self,
        message: str,
        mode: str = "deliberate",
    ):
        """Stream chat response (blocking generator)."""
        loop = self._get_loop()

        async def _stream():
            self._async_client.stream({
                "message": message,
                "mode": mode,
            })
            async for chunk in self._async_client.stream({"message": message, "mode": mode}):
                yield chunk

        return loop.run_until_complete(self._async_generator(loop, _stream()))

    def _async_generator(self, loop, gen):
        """Convert async generator to sync generator."""
        while True:
            try:
                item = loop.run_until_complete(gen.__anext__())
                yield item
            except StopAsyncIteration:
                break

    def __enter__(self):
        self.connect()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.disconnect()